client/src/legacy/LeftAndMain.EditForm.js

```diff
@@ -110,28 +110,91 @@ $.entwine('ss', function($){
     },
     'from .cms-tabset': {
       onafterredrawtabs: function () {
-        // Show validation errors if necessary
-        if(this.hasClass('validationerror')) {
-          // Ensure the first validation error is visible
-          var tabError = this.find('.message.validation, .message.required').first().closest('.tab');
-          $('.cms-container').clearCurrentTabState(); // clear state to avoid override later on
-
-          // Attempt #1: Look for nearest .ss-tabset (usually nested deeper underneath a .cms-tabset).
-          var $tabSet = tabError.closest('.ss-tabset');
-
-          // Attempt #2: Next level in tab-ception, try to select the tab within this higher level .cms-tabset if possible
-          if (!$tabSet.length) {
-            $tabSet = tabError.closest('.cms-tabset');
-          }
 
-          if ($tabSet.length) {
-            $tabSet.tabs('option', 'active', tabError.index('.tab'));
-          } else if (!this.getValidationErrorShown()) {
-            // Ensure that this error message popup won't be added more than once
-            this.setValidationErrorShown(true);
-            errorMessage(ss.i18n._t('Admin.VALIDATIONERROR', 'Validation Error'));
-          }
+        // This function will:
+        // - Add an invalid icon on each tab that contains form fields than failed validation
+        // - Set an alert message in the edit form error banner
+        const iconClass = 'font-icon-attention-1 tab-attention';
+        const iconTitle = ss.i18n._t(
+          'VALIDATION_ERRORS_IN_TAB',
+          'This tab contains validation errors.'
+        );
+        const alertMessageText = ss.i18n._t(
+          'Admin.VALIDATION_ERRORS_ON_PAGE',
+          'There are validation errors on this page, please fix them before saving or publishing.'
+        );
+        const toastNotificationMessage = ss.i18n._t(
+          'Admin.VALIDATIONERROR',
+          'Validation Error'
+        );
+        const $editFormErrorBanner = $("#Form_EditForm_error");
+
+        // Remove any existing invalid tab icons and screen-reader text
+        this.find('.tab-attention, .tab-validation-error-sr').remove();
+
+        // Check if there are any form validation errors
+        let validationErrorExists = false;
+
+        // Validation errors from DataObject::getCMSValidator()
+        if (this.hasClass('validationerror')) {
+          validationErrorExists = true;
+        }
+
+        // Validation errors from DataObject::validate() .. ValidationResult::addError();
+        if ($editFormErrorBanner.html() !== '') {
+          validationErrorExists = true;
+        }
+
+        // Validation errors from DataObject::validate() .. ValidationResult::addFieldError()
+        if (this.find('.alert.error[@role="alert"]').length > 0) {
```

Comment:
Is it safe to drop the role attribute part of this selector? I ask because these errors shouldn't even _have_ that attribute, and it would be best to keep it out of the selector if possible. You could instead add some surrounding context like the field holder in to make it more precise.

Reply:
Yup it's safe, removed it